edgedb/ai/core.py

```diff
@@ -0,0 +1,174 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2024-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+import typing
+
+import edgedb
+import httpx
+import httpx_sse
+
+from . import types
+
+
+def create_AI(client: edgedb.Client, **kwargs) -> EdgeDBAI:
+    client.ensure_connected()
+    return EdgeDBAI(client, types.AIOptions(**kwargs))
+
+
+async def create_async_AI(
+    client: edgedb.AsyncIOClient, **kwargs
+) -> AsyncEdgeDBAI:
+    await client.ensure_connected()
+    return AsyncEdgeDBAI(client, types.AIOptions(**kwargs))
+
+
+class BaseEdgeDBAI:
+    options: types.AIOptions
+    context: types.QueryContext
+    client_cls = NotImplemented
+
+    def __init__(
+        self,
+        client: typing.Union[edgedb.Client, edgedb.AsyncIOClient],
+        options: types.AIOptions,
+        **kwargs,
+    ):
+        pool = client._impl
+        host, port = pool._working_addr
+        params = pool._working_params
+        proto = "http" if params.tls_security == "insecure" else "https"
+        branch = params.branch
+        self.options = options
+        self.context = types.QueryContext(**kwargs)
+        args = dict(
+            base_url=f"{proto}://{host}:{port}/branch/{branch}/ext/ai",
+            verify=params.ssl_ctx,
+        )
+        if params.password is not None:
+            args["auth"] = (params.user, params.password)
+        elif params.secret_key is not None:
+            args["headers"] = {"Authorization": f"Bearer {params.secret_key}"}
+        self._init_client(**args)
+
+    def _init_client(self, **kwargs):
+        raise NotImplementedError
+
+    def with_config(self, **kwargs) -> typing.Self:
+        cls = type(self)
+        rv = cls.__new__(cls)
+        rv.options = self.options.derive(kwargs)
+        rv.context = self.context
+        rv.client = self.client
+        return rv
+
+    def with_context(self, **kwargs) -> typing.Self:
+        cls = type(self)
+        rv = cls.__new__(cls)
+        rv.options = self.options
+        rv.context = self.context.derive(kwargs)
+        rv.client = self.client
+        return rv
+
+
+class EdgeDBAI(BaseEdgeDBAI):
+    client: httpx.Client
+
+    def _init_client(self, **kwargs):
+        self.client = httpx.Client(**kwargs)
+
```

Comment:
Need a wrapper for the `/embeddings` endpoint too.

Reply:
I'll add it in a new PR 🙏